### PR TITLE
feat: add HardSourceWebpackPlugin

### LINKED ISSUE #4107: Add HardSourceWebpackPlugin
### What problem does this feature solve?

Might be worth a try to improve speed/DX: https://github.com/mzgoddard/hard-source-webpack-plugin

<div align="right"><sub><em>This feature request is available on <a href="https://cmty.app/nuxt">Nuxt</a> community (<a href="https://cmty.app/nuxt/nuxt.js/issues/c7966">#c7966</a>)</em></sub></div>

### PR DESCRIPTION
**Pull request by TheAlexLichter**
Adds experimental support for https://github.com/mzgoddard/hard-source-webpack-plugin (disabled by default)

## Types of changes

- [ ] Bug fix (a non-breaking change which fixes an issue)
- [x] New feature (a non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Description

Resolves https://github.com/nuxt/nuxt.js/issues/4107

## Checklist:

- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly. (PR: #)
- [ ] I have added tests to cover my changes (if not applicable, please state why)
- [x] All new and existing tests are passing.

### CODE REVIEW

**Review by pi0 — APPROVED** [review]

File: packages/webpack/src/config/base.js [hunk 1]
```diff
@@ -6,6 +6,7 @@ import cloneDeep from 'lodash/cloneDeep'
 import escapeRegExp from 'lodash/escapeRegExp'
 import VueLoader from 'vue-loader'
 import ExtractCssChunks from 'extract-css-chunks-webpack-plugin'
+import HardSource from 'hard-source-webpack-plugin'
```
Comment (pi0) [inline packages/webpack/src/config/base.js]: For naming convention, i suggest importing as `HardSourcePlugin`

**Review by pi0 — DISMISSED** [review]

**Comment by aldarund** [comment]
Heh, i wanted to add it too since its have fixed issues with recent webpack recently . but here u are :)

**Review by aldarund — DISMISSED** [review]

File: packages/webpack/src/config/base.js (hunk 1 above)
Comment (manniL) [reply packages/webpack/src/config/base.js]: Can do that, though we did not do the same iwth with `ExtractCssChunks`.

**Comment by manniL** [comment]
@aldarund Sorry :see_no_evil:

File: packages/webpack/src/config/base.js (hunk 1 above)
Comment (pi0) [reply packages/webpack/src/config/base.js]: But we do it for `TerserWebpackPlugin` :D Feel free using whatever you think is better. The point with convention is consistency.

Comment (pi0) [reply packages/webpack/src/config/base.js]: Also most of the `client.js` imports are in `*Plugin` form: https://github.com/nuxt/nuxt.js/blob/c3db15683a349d2a2d43d9c9965c925b401df764/packages/webpack/src/config/client.js#L8

**Review by clarkdo — DISMISSED** [review]

**Comment by manniL** [comment]
Done :+1:

**Pull request merged**